patches/tModLoader/Terraria.ModLoader/DoubleJumpHooks.cs

```diff
@@ -0,0 +1,195 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+
+namespace Terraria.ModLoader
+{
+	/// <summary>
+	/// This is where all DoubleJumpHooks hooks are gathered and called.
+	/// </summary>
+	public static class DoubleJumpHooks
+	{
+		private static readonly IList<ModDoubleJump> doubleJumps = new List<ModDoubleJump>();
+
+		/// <summary>
+		/// This is the index of the jump that will be executed next
+		/// </summary>
+		//It's okay to have it static here because all the jumping logic happens player-by-player which is reset to -1 when the next player jumping logic happens.
+		//Also to reduce patch size having it here instead of as a local variable in Player.cs
+		private static int nextJump = -1;
+
+		private class HookList
+		{
+			public int[] arr = new int[0];
+			public readonly MethodInfo method;
+
+			public HookList(MethodInfo method)
+			{
+				this.method = method;
+			}
+		}
+
+		private static List<HookList> hooks = new List<HookList>();
+
+		private static HookList AddHook<F>(Expression<Func<ModDoubleJump, F>> func)
+		{
+			var hook = new HookList(ModLoader.Method(func));
+			hooks.Add(hook);
+			return hook;
+		}
+
+		internal static void Add(ModDoubleJump doubleJump)
+		{
+			doubleJump.index = doubleJumps.Count;
+			doubleJumps.Add(doubleJump);
+		}
+
+		internal static void RebuildHooks()
+		{
+			foreach (var hook in hooks) {
+				hook.arr = ModLoader.BuildGlobalHook(doubleJumps, hook.method).Select(p => p.index).ToArray();
+			}
+		}
+
+		internal static void Unload()
+		{
+			doubleJumps.Clear();
+		}
+
+		internal static void SetupPlayer(Player player)
+		{
+			player.modDoubleJumps = doubleJumps.Select(doubleJump => doubleJump.CreateFor(player)).ToArray();
+		}
+
+		private static HookList HookMidJump = AddHook<Action>(d => d.MidJump);
+
+		/// <summary>
+		/// Allows you to create visuals while the jump is happening
+		/// </summary>
+		/// <param name="player">Jumping player</param>
+		public static void MidJump(Player player)
+		{
+			foreach (int index in HookMidJump.arr) {
+				ModDoubleJump doubleJump = player.modDoubleJumps[index];
+				if (doubleJump.IsMidJump) {
+					player.modDoubleJumps[index].MidJump();
+					return; // Only one jump active at a time
+				}
+			}
+		}
+
+		/// <summary>
+		/// Do the jump if a valid <see cref="nextJump"/> is found in SetNextJump
+		/// </summary>
+		/// <param name="player">Jumping player</param>
+		/// <param name="playSound">if left on true, will play the default double jump sound</param>
+		/// <returns>Jump height multiplier</returns>
+		internal static float Jump(Player player, ref bool playSound)
+		{
+			if (nextJump != -1) {
```

Comment:
Yeah not sure if I like this design. It makes more sense to just store an `activeJump` with ModJump type, and have the changing logic happen during keypress switches/changes. Then in the rest of the code, you just call `activeJump.logic()` etc. Reduces 'magic numbers' happening

Reply:
Not sure I'm following, please check the latest commit

Reply:
Yup. See how much simpler it is and how much code you're able to throw away? :-)